test/e2e_node/garbage_collector_test.go

```diff
@@ -0,0 +1,153 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_node
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	dockertypes "github.com/docker/engine-api/types"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	docker "k8s.io/kubernetes/pkg/kubelet/dockertools"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+/*
+GarbageCollect tests that the Kubelet conforms to the Kubelet Garbage Collection Policy, found here:
+http://kubernetes.io/docs/admin/garbage-collection/
+TODO (dashpole): Once the Container Runtime Interface (CRI) is complete, generalize to other runtimes (other than docker)
+TODO (dashpole): Once dynamic config is possible, test different values for maxPerPodContainer and maxContainers
+*/
+const (
+	defaultDockerEndpoint                = "unix:///var/run/docker.sock"
+	defaultRuntimeRequestTimeoutDuration = 2 * time.Minute
+	maxPerPodContainer                   = 2
+	maxTotalContainers                   = -1
+
+	runtimePollInterval = 10 * time.Second
+	deleteTimeout       = 2 * time.Minute
+
+	containerNamePrefix = "gc-test-container"
+)
+
+var _ = framework.KubeDescribe("GarbageCollect [Serial]", func() {
+	f := framework.NewDefaultFramework("garbage-collect-test")
+	runtime := docker.ConnectToDockerOrDie(defaultDockerEndpoint, defaultRuntimeRequestTimeoutDuration)
+
+	Context("", func() {
+		AfterEach(func() {
+			glog.Infof("Summary of node events during the garbage collection test:")
+			err := framework.ListNamespaceEvents(f.ClientSet, f.Namespace.Name)
+			framework.ExpectNoError(err)
+			glog.Infof("Summary of pod events during the garbage collection test:")
+			err = framework.ListNamespaceEvents(f.ClientSet, "")
+			framework.ExpectNoError(err)
+		})
+		maxPerPodContainerGCTest(f, runtime, 6, 1)
+		maxPerPodContainerGCTest(f, runtime, 2, 4)
+		maxPerPodContainerGCTest(f, runtime, 3, 6)
+		maxPerPodContainerGCTest(f, runtime, 4, 1)
+	})
+
+})
+
+// Tests that maxPerPodContainer limits the number of non-garbage collected containers for each container created.
+func maxPerPodContainerGCTest(f *framework.Framework, runtime docker.DockerInterface, restartCount int, numContainers int) {
+	Context(fmt.Sprintf("when we create and delete pods with containers %v times", restartCount), func() {
+		podName := fmt.Sprintf("gc-test-pod%d.%d", numContainers, restartCount)
+		BeforeEach(func() {
+			f.PodClient().CreateSync(getPodWithRestartCount(podName, numContainers, restartCount))
+			time.Sleep(10 * time.Second)
```

Comment:
Hm, it makes sense. If so, 10 second may not be enough, because we have restart backoff. The backoff period is 10s, 20s, 40s, 80s... Cap to 300s.

There are several ways to check whether the containers go into stable state:
1) Use restart count. There is a field called `RestartCount` in container status, we can wait until the restart count reaches the expected restart count. The cons is that if garbage collection logic is wrong, it may remove the "latest container", in that case restart count will be reset because the lack of historical data.

2) Use container log. `echo "hello" | tee file_name`, so that we can check container log to see whether the output reaches expected restart count.

3) Use a another container mount the volume, and only exits when there are expected lines in the file.

Any is ok. 1) should be simplest.